### PR TITLE
Assylum

### PR DESCRIPTION
**Pull request by Tweeticoats**
Adding Assylum.com scraper

### CODE REVIEW

File: scrapers/Assylum.yml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        Name:
+          selector: //p[@class='tags']/a/text()
+      Details: //p[@class='mas_longdescription']/text()
+      Date:
+        selector: //span[@class='lc_info mas_description']/text()[2]
+        parseDate: January 2, 2006
+debug:
+  printHTML: true
+
+# Last Updated August 26, 2020
```
Comment (halorrr) [inline scrapers/Assylum.yml:36]: Date needs to be updated

File: scrapers/Assylum.yml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+      Tags:
+        Name:
+          selector: //p[@class='tags']/a/text()
+      Details: //p[@class='mas_longdescription']/text()
+      Date:
+        selector: //span[@class='lc_info mas_description']/text()[2]
+        parseDate: January 2, 2006
+debug:
+  printHTML: true
```
Comment (halorrr) [inline scrapers/Assylum.yml:34]: Need to clean up debug

File: scrapers/Assylum.yml [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    scene:
+      Title:
+        selector: //h3[@class='mas_title']/text()
+      Performers:
+        Name: 
+          selector: //span[@class='lc_info mas_description']/text()[1]
+          postProcess:
+            - replace:
+              - regex: (.+)\,
+                with: $1
```
Comment (halorrr) [inline scrapers/Assylum.yml:18]: Scenes like this one `https://www.assylum.com/session/Noemie_Bilas/ass_eating_domination/603` the Performer name I got was `Noemie Bilas, Day 1` which could be fixed with some more regex.

But scenes like this `https://www.assylum.com/session/Leah-Winters/Sadistic-Sustenance/Rough-anal-domination-and-extreme-ass-fucking/590` don't have the performer names in the title. Not sure there is anything that can be done about that as I couldn't find any reliable way to get a performer name here.

File: scrapers/Assylum.yml [hunk 4]
```diff
@@ -0,0 +1,36 @@
+name: assylum
+sceneByURL:
+  - action: scrapeXPath
+    url:
+      - https://www.assylum.com/session/
```
Comment (halorrr) [inline scrapers/Assylum.yml:5]: Protocol and www can be dropped for just `assylum.com/session/`

**Comment by halorrr** [comment]
Based on my comments this would be the updated scraper:

```YAML
name: assylum
sceneByURL:
  - action: scrapeXPath
    url:
      - assylum.com/session/
    scraper: sceneScraper
xPathScrapers:
  sceneScraper:
    scene:
      Title:
        selector: //h3[@class='mas_title']/text()
      Performers:
        Name:
          selector: //span[@class='lc_info mas_description']/text()[1]
          postProcess:
            - replace:
              - regex: (.+)\,
                with: $1
      Image:
        selector: //div[@class='mainpic']/comment()
        postProcess:
          - replace:
            - regex: .*src="(.*?)".*
              with:  https://www.assylum.com/$1

      Tags:
        Name:
          selector: //p[@class='tags']/a/text()
      Details: //p[@class='mas_longdescription']/text()
      Date:
        selector: //span[@class='lc_info mas_description']/text()[2]
        parseDate: January 2, 2006

# Last Updated October 23, 2020
```

**Comment by bnkai** [comment]
i think we can also add the studio
```yaml
      Studio:
        Name:
          fixed: Assylum
```

@halorrr for the performers aside from the missing performer there is also the case with multiple performers and mixed performers and no performers in the description span
eg
```
https://www.assylum.com/session/Leah-Winters/Sadistic-Sustenance/Rough-anal-domination-and-extreme-ass-fucking/590
https://www.assylum.com/session/Noemie_Bilas/ass_eating_domination/603
https://www.assylum.com/session/Anastasia-Rose-vs-Lydia-Black/hard-anal-domination-with-rough-sex-and-piss/547
https://www.assylum.com/session/Rebel-Rhyder/Order-of-the-Red-Star/Extreme-Anal-Gangbang-with-hard-domination/552
https://www.assylum.com/session/leah-winter/quarantin-dreams-1/submission-inc/hard-anal-domination/589
https://www.assylum.com/session/Lexi-Grey/Anal-Crazy-School/Cruel-ass-fucking-with-ass-to-mouth-and-piss-drinking-domination/569
```

with this 
```yaml
      Performers:
        Name:
          selector: //span[@class='lc_info mas_description']/text()[1]
          postProcess:
            - replace:
                - regex: "([^\\,]*[^a-zA-Z\\s\\,]+[a-zA-Z\\s]*)\\," # performer names shouldnt have any chars other than spaces and  letters
                  with:
                - regex: \,\s+
                  with: ","
                - regex: ^\s+
                  with:
          split: ","
```
we can clear most non performers from the results except from cases like the last url where there is no way of knowing if its a performer name or not. Some more tweaking might be necessary.

**Comment by halorrr** [comment]
@bnkai I double checked the model list on the site and currently `performer names shouldnt have any chars other than spaces and  letters` is a safe assumption but may not be future proof. Some examples that come to mind:

- `LaSirena69` - _Currently growing in popularity and the only one I know of with a number in the name_
- `J.C. Simpson` - _Depending on the site, sometimes she drops the periods, but that isn't a sure thing_ (There are a couple others that exist like her though)

I could have sworn I've seen performers with hyphens in their names, but couldn't find an example so maybe I'm wrong on that one. For sure numbers and periods should be accounted for though if we want to think future proofing on this one.

**Comment by bnkai** [comment]
@halorrr we only care about performers on the specific site for this scraper. Adding a number or a period for the performers in this site is probably increasing the probabililty for false positives imho. I checked the models pages (only 2) and the models don't seem to have any extra number or period or special char to their name. Can you think of any better regex for cleaning the performers selector?

**Comment by halorrr** [comment]
@bnkai I have created a monster regex but it seems to work in all edge cases (including the future proofed ones I mentioned). Let me know if you want me to break it down piece by piece cause it got a little crazy.

```YAML
name: assylum
sceneByURL:
  - action: scrapeXPath
    url:
      - assylum.com/session/
    scraper: sceneScraper
xPathScrapers:
  sceneScraper:
    scene:
      Title:
        selector: //h3[@class='mas_title']/text()
      Performers:
        Name:
          selector: //span[@class='lc_info mas_description']/text()[1]
          postProcess:
            - replace:
                - regex: (?:[\w\s]+\:\s)?([A-Z]?[a-z]+\d?\d?\s[A-Z]?[a-z]+\d?\d?,|[A-Z]?[a-z]+\d?\d?,)\s?([A-Z]?[a-z]+\d?\d?\s[A-Z]?[a-z]+\d?\d?,|[A-Z]?[a-z]+\d?\d?,)?\s?([A-Z]?[a-z]+\d?\d?\s[A-Z]?[a-z]+\d?\d?,|[A-Z]?[a-z]+\d?\d?,)?|.+
                  with: $1$2$3
                - regex: \,\s+
                  with: ","
                - regex: ^\s+
                  with:
          split: ","
      Image:
        selector: //div[@class='mainpic']/comment()
        postProcess:
          - replace:
            - regex: .*src="(.*?)".*
              with:  https://www.assylum.com/$1

      Tags:
        Name:
          selector: //p[@class='tags']/a/text()
      Details: //p[@class='mas_longdescription']/text()
      Date:
        selector: //span[@class='lc_info mas_description']/text()[2]
        parseDate: January 2, 2006
      Studio:
        Name:
          fixed: Assylum

# Last Updated November 4, 2020
```

**Pull request merged**